File: vbnet/anagram/AnagramTests.vb
Imports XUnit


Public Class AnagramTest
    <Fact>
    Public Sub NoMatches()
        Dim detector = New Anagram("diaper")
        Dim words = New String() {"hello", "world", "zombies", "pants"}
        Dim results = New String() {}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub DetectSimpleAnagram()
        Dim detector = New Anagram("ant")
        Dim words = New String() {"tan", "stand", "at"}
        Dim results = New String() {"tan"}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub DetectMultipleAnagrams()
        Dim detector = New Anagram("master")
        Dim words = New String() {"stream", "pigeon", "maters"}
        Dim results = New String() {"maters", "stream"}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub DoesNotConfuseDifferentDuplicates()
        Dim detector = New Anagram("galea")
        Dim words = New String() {"eagle"}
        Dim results = New String() {}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub IdenticalWordIsNotAnagram()
        Dim detector = New Anagram("corn")
        Dim words = New String() {"corn", "dark", "Corn", "rank", "CORN", "cron",
            "park"}
        Dim results = New String() {"cron"}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub EliminateAnagramsWithSameChecksum()
        Dim detector = New Anagram("mass")
        Dim words = New String() {"last"}
        Dim results = New String(-1) {}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub EliminateAnagramSubsets()
        Dim detector = New Anagram("good")
        Dim words = New String() {"dog", "goody"}
        Dim results = New String(-1) {}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub DetectAnagrams()
        Dim detector = New Anagram("allergy")
        Dim words = New String() {"gallery", "ballerina", "regally", "clergy", "largely", "leading"}
        Dim results = New String() {"gallery", "largely", "regally"}
        Assert.Equal(detector.Match(words), results)
    End Sub

    <Fact>
    Public Sub AnagramsAreCaseInsensitive()
        Dim detector = New Anagram("Orchestra")
        Dim words = New String() {"cashregister", "Carthorse", "radishes"}
        Dim results = New String() {"Carthorse"}
        Assert.Equal(detector.Match(words), results)
    End Sub

End Class

File: swift/anagram/Tests/AnagramTests/AnagramTests.swift
import XCTest

@testable import Anagram

class AnagramTests: XCTestCase {
  let runAll = Bool(ProcessInfo.processInfo.environment["RUNALL", default: "false"]) ?? false

  func testNoMatches() {
    let anagram = Anagram(word: "diaper")
    let results = anagram.match(["hello", "world", "zombies", "pants"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testDetectsTwoAnagrams() throws {
    let anagram = Anagram(word: "solemn")
    let results = anagram.match(["lemons", "cherry", "melons"])
    let expected = ["lemons", "melons"]
    XCTAssertEqual(results, expected)
  }

  func testDoesNotDetectAnagramSubsets() throws {
    let anagram = Anagram(word: "good")
    let results = anagram.match(["dog", "goody"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testDetectsAnagram() throws {
    let anagram = Anagram(word: "listen")
    let results = anagram.match(["enlists", "google", "inlets", "banana"])
    let expected = ["inlets"]
    XCTAssertEqual(results, expected)
  }

  func testDetectsThreeAnagrams() throws {
    let anagram = Anagram(word: "allergy")
    let results = anagram.match(["gallery", "ballerina", "regally", "clergy", "largely", "leading"])
    let expected = ["gallery", "regally", "largely"]
    XCTAssertEqual(results, expected)
  }

  func testDetectsMultipleAnagramsWithDifferentCase() throws {
    let anagram = Anagram(word: "nose")
    let results = anagram.match(["Eons", "ONES"])
    let expected = ["Eons", "ONES"]
    XCTAssertEqual(results, expected)
  }

  func testDoesNotDetectNonAnagramsWithIdenticalChecksum() throws {
    let anagram = Anagram(word: "mass")
    let results = anagram.match(["last"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testDetectsAnagramsCaseInsensitively() throws {
    let anagram = Anagram(word: "Orchestra")
    let results = anagram.match(["cashregister", "Carthorse", "radishes"])
    let expected = ["Carthorse"]
    XCTAssertEqual(results, expected)
  }

  func testDetectsAnagramsUsingCaseInsensitiveSubject() throws {
    let anagram = Anagram(word: "Orchestra")
    let results = anagram.match(["cashregister", "carthorse", "radishes"])
    let expected = ["carthorse"]
    XCTAssertEqual(results, expected)
  }

  func testDetectsAnagramsUsingCaseInsensitivePossibleMatches() throws {
    let anagram = Anagram(word: "orchestra")
    let results = anagram.match(["cashregister", "Carthorse", "radishes"])
    let expected = ["Carthorse"]
    XCTAssertEqual(results, expected)
  }

  func testDoesNotDetectAnAnagramIfTheOriginalWordIsRepeated() throws {
    let anagram = Anagram(word: "go")
    let results = anagram.match(["go Go GO"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testAnagramsMustUseAllLettersExactlyOnce() throws {
    let anagram = Anagram(word: "tapper")
    let results = anagram.match(["patter"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testWordsAreNotAnagramsOfThemselves() throws {
    let anagram = Anagram(word: "BANANA")
    let results = anagram.match(["BANANA"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testWordsAreNotAnagramsOfThemselvesEvenIfLetterCaseIsPartiallyDifferent() throws {
    let anagram = Anagram(word: "BANANA")
    let results = anagram.match(["Banana"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testWordsAreNotAnagramsOfThemselvesEvenIfLetterCaseIsCompletelyDifferent() throws {
    let anagram = Anagram(word: "BANANA")
    let results = anagram.match(["banana"])
    let expected = [String]()
    XCTAssertEqual(results, expected)
  }

  func testWordsOtherThanThemselvesCanBeAnagrams() throws {
    let anagram = Anagram(word: "LISTEN")
    let results = anagram.match(["LISTEN", "Silent"])
    let expected = ["Silent"]
    XCTAssertEqual(results, expected)
  }
}
